network/p2p/inspector/validation/control_message_validation_inspector_test.go

```diff
@@ -1,376 +1,621 @@
+		suite.rpcTracker.On("WasIHaveRPCSent", mock.AnythingOfType("string")).Return(true).Maybe()
+		suite.distributor.On("Distribute", mock.AnythingOfType("*p2p.InvCtrlMsgNotif")).Return(nil).Maybe()
+		suite.config.IWantRPCInspectionConfig.MaxMessageIDSampleSize = 100
+		suite.inspector.Start(suite.signalerCtx)
+
 		iWantsGreaterThanMaxSampleSize := unittest.P2PRPCFixture(unittest.WithIWants(unittest.P2PRPCIWantFixtures(10, 200)...))
 		iWantsLessThanMaxSampleSize := unittest.P2PRPCFixture(unittest.WithIWants(unittest.P2PRPCIWantFixtures(10, 50)...))
-		peerID := peer.ID("peer")
-		rpcTracker.On("LastHighestIHaveRPCSize").Return(int64(100)).Twice()
-		inspector.truncateIWantMessages(peerID, iWantsGreaterThanMaxSampleSize)
-		inspector.truncateIWantMessages(peerID, iWantsLessThanMaxSampleSize)
-		for _, iWant := range iWantsGreaterThanMaxSampleSize.GetControl().GetIwant() {
-			// rpc with iWants message ids greater than configured max sample size should be truncated to MaxSampleSize
-			require.Len(t, iWant.GetMessageIDs(), inspector.config.IWantRPCInspectionConfig.MaxMessageIDSampleSize)
-		}
-		for _, iWant := range iWantsLessThanMaxSampleSize.GetControl().GetIwant() {
-			// rpc with iWants less than MaxSampleSize should not be truncated
-			require.Len(t, iWant.GetMessageIDs(), 50)
-		}
+
+		from := unittest.PeerIdFixture(t)
+		require.NoError(t, suite.inspector.Inspect(from, iWantsGreaterThanMaxSampleSize))
+		require.NoError(t, suite.inspector.Inspect(from, iWantsLessThanMaxSampleSize))
+		require.Eventually(t, func() bool {
+			for _, iWant := range iWantsGreaterThanMaxSampleSize.GetControl().GetIwant() {
+				// rpc with iWants message ids greater than configured max sample size should be truncated to MaxSampleSize
+				if len(iWant.GetMessageIDs()) != suite.config.IWantRPCInspectionConfig.MaxMessageIDSampleSize {
+					return false
+				}
+			}
+			for _, iWant := range iWantsLessThanMaxSampleSize.GetControl().GetIwant() {
+				// rpc with iWants less than MaxSampleSize should not be truncated
+				if len(iWant.GetMessageIDs()) != 50 {
+					return false
+				}
+			}
+			return true
+		}, time.Second, 500*time.Millisecond)
 	})
 }
 
 // TestControlMessageValidationInspector_processInspectRPCReq verifies the correct behavior of control message validation.
 // It ensures that valid RPC control messages do not trigger erroneous invalid control message notifications,
 // while all types of invalid control messages trigger expected notifications.
-func TestControlMessageValidationInspector_processInspectRPCReq(t *testing.T) {
-	t.Run("processInspectRPCReq should not disseminate any invalid notification errors for valid RPC's", func(t *testing.T) {
-		inspector, distributor, rpcTracker, _, sporkID := inspectorFixture(t)
-		defer distributor.AssertNotCalled(t, "Distribute")
+func (suite *ControlMsgValidationInspectorSuite) TestControlMessageValidationInspector_processInspectRPCReq() {
```

Comment:
Why is this test function full of sub-tests that each manually call `suite.SetupTest()`? Why not have each sub-test a stand alone test of the test suite instead of a sub-test?

Reply:
Each Test case is made up of related sub-test cases, it's nice to organize them based on this grouping. You can utilize the suite pattern with sub test that run using t.Run by manually calling SetupTest in each subtest.